library/src/main/java/com/esafirm/rxdownloader/RxDownloader.java

```diff
@@ -47,44 +48,108 @@ private DownloadManager getDownloadManager() {
+                        return progressState.path;
+                    }
+                })
+                .firstOrError();
+    }
 
+    public Observable<DownloadState> downloadWithProgress(DownloadManager.Request request) {
+        final long downloadId = getDownloadManager().enqueue(request);
+
+        final PublishSubject<DownloadState> publishSubject = PublishSubject.create();
+        progressSubjectMap.put(downloadId, publishSubject);
+
+        Observable<DownloadState> observable =
+                Observable.create(new ObservableOnSubscribe<DownloadState>() {
+                    @Override
+                    public void subscribe(ObservableEmitter<DownloadState> e) throws Exception {
+                        DownloadManager.Query query = new DownloadManager.Query();
+                        query.setFilterById(downloadId);
+                        while (true) {
+                            Cursor cursor = getDownloadManager().query(query);
+                            if (!cursor.moveToFirst()) {
+                                cursor.close();
+                                downloadManager.remove(downloadId);
+                                publishSubject.onError(new IllegalStateException("Cursor empty, this shouldn't happen"));
+                                progressSubjectMap.remove(downloadId);
+                                return;
+                            }
+                            int bytesDownloaded = cursor.getInt(cursor
+                                    .getColumnIndex(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
+                            int bytesTotal = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));
+
+                            int status = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_STATUS));
+                            int uriIndex = cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI);
+                            String downloadedPackageUriString = cursor.getString(uriIndex);
+                            cursor.close();
+                            if (status == DownloadManager.STATUS_SUCCESSFUL) {
+                                e.onNext(DownloadState.create(100, downloadedPackageUriString));
+                                e.onComplete();
+                                progressSubjectMap.remove(downloadId);
+                                break;
+                            } else if (status == DownloadManager.STATUS_FAILED) {
+                                e.onError(new RuntimeException("Download not complete"));
+                                progressSubjectMap.remove(downloadId);
+                                break;
+                            }
+                            final int progress = (int) ((bytesDownloaded * 1f) / bytesTotal * 100);
+                            e.onNext(DownloadState.create(progress, null));
+                            Thread.sleep(PROGRESS_INTERVAL_MILLIS);
```

Comment:
I think, instead of using `Thread.sleep` directly we could leverage `Observable.timer`

Reply:
Ah right. Yeah, that makes sense.

Reply:
Am trying to do this but not sure how to use `interval` with `create`. Any ideas?

Reply:
Like this? https://gist.github.com/Kisty/4dff6c4ba428b611981ba76ffd79af80